catalog/app/containers/Bucket/HandleNoSlashDir.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react'
+import * as RRDom from 'react-router-dom'
+
+import Placeholder from 'components/Placeholder'
+import * as Model from 'model'
+import * as AWS from 'utils/AWS'
+import * as NamedRoutes from 'utils/NamedRoutes'
+import * as s3paths from 'utils/s3paths'
+
+import * as requests from './requests'
+import { displayError } from './errors'
+
+// If object exists, then this is 100% an object page
+function useIsObject(handle: Model.S3.S3ObjectLocation) {
+  const [exists, setExists] = React.useState(null)
+
+  const s3 = AWS.S3.use()
+  React.useEffect(() => {
+    const { bucket, key, version } = handle
+    function resolveObjectExistence() {
+      requests
+        .getObjectExistence({
+          s3,
+          bucket,
+          key,
+          version,
+        })
+        .then(
+          requests.ObjectExistence.case({
+            Exists: () => true,
+            _: () => false,
+          }),
+        )
+        .then(setExists)
+    }
+    resolveObjectExistence()
+  }, [s3, handle])
+
+  return exists
+}
+
+// If prefix contains at least something, then it is a directory.
+// S3 can not have empty directories, because directories are virtual,
+//    they are based on paths for existing keys (file)
+function useIsDirectory(handle: Model.S3.S3ObjectLocation) {
+  const bucketListing = requests.useBucketListing()
+  return React.useCallback(async () => {
+    const { bucket, key } = handle
+    const path = s3paths.ensureSlash(key)
+    const { dirs, files } = await bucketListing({ bucket, path, maxKeys: 1 })
+    return !!dirs.length || !!files.length
+  }, [bucketListing, handle])
+}
+
+function useHandleNoSlashDir(
+  handle: Model.S3.S3ObjectLocation,
+): 'loading' | 'dir' | 'file' | Error {
+  const isObject = useIsObject(handle)
+  const requestIsDirectory = useIsDirectory(handle)
+
+  const [pageType, setPageType] = React.useState<'loading' | 'dir' | 'file' | Error>(
+    'loading',
+  )
+
+  React.useEffect(() => {
+    function resolvePageType() {
+      if (isObject === null) return
+
+      if (isObject) {
+        setPageType('file')
+        return
+      }
+
+      requestIsDirectory()
+        .then((isDir) => setPageType(isDir ? 'dir' : 'file'))
+        .catch((e) => setPageType(e instanceof Error ? e : new Error(`${e}`)))
+    }
+    resolvePageType()
+  }, [handle, isObject, requestIsDirectory])
+
+  return pageType
```

Comment:
using state here is error-prone, because you have to catch state transitions and update computed state.
maybe try something like this?

```suggestion
  const isObject = useIsObject(handle)
  const isDirectory = useIsDirectory(handle)

  if (isObject === null) return 'loading'
  if (isObject) return 'file'
  if (isDirectory === null) return 'loading'
  return isDirectory ? 'dir' : 'file'
  // TODO: error handling
```

unlike the original version, it will always make the second (bucketListing) call, but we won't be waiting for it, so performance-wise it should be the same.
you'd have to modify `useIsDirectory` to encapsulate request state similar to `useIsObject`.
to avoid the second request, you can add a parameter (say, `pause` or smth) to `useIsDirectory` to make the request conditional.